cpp/include/cugraph/algorithms.hpp

```diff
@@ -1606,6 +1606,31 @@ void core_number(raft::handle_t const& handle,
                  size_t k_last           = std::numeric_limits<size_t>::max(),
                  bool do_expensive_check = false);
 
+/**
+ * @brief   Extract K Core of a graph
+ *
+ * @throws     cugraph::logic_error when an error occurs.
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @tparam edge_t Type of edge identifiers. Needs to be an integral type.
+ * @tparam weight_t Type of edge weights. Needs to be a floating point type.
+ * @tparam multi_gpu Flag indicating whether template instantiation should target single-GPU (false)
+ * @param  graph           cuGraph graph in coordinate format
+ * @param  k               Order of the core. This value must not be negative.
+ * @param  core_numbers    Output from core_number algorithm
+ *
+ * @return edge list for the graph
+ */
+template <typename vertex_t, typename edge_t, typename weight_t, bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           std::optional<rmm::device_uvector<weight_t>>>
+k_core(raft::handle_t const& handle,
```

Comment:
Something to think about...

So, we're passing core numbers here. So in this case, should we really have both this function and induced subgraph?

seeing core numbers, we can easily extract a vertex list in a single thrust call (e.g. thrust::copy_if). Then, we can pass a vertex list to induced subgraph.

Not sure this function adds enough convenience to justify increase in compile time/binary size.

Reply:
Maybe OK, if this calls explicitly instantiated induced_subgraph, increases in compile time/binary size might be minimal...

Still debating between whether this function should take core_numbers or just call core_numbers internally.... (to maximize user convenience, if not, a user can just separately call core_numbers, something like thurst::copy_if, and induced_subgraph).

Reply:
I had that debate internally.  Ultimately I concluded that if we're going to expose a complete k-core implementation at the python level we should probably expose it at the lower levels as well.  It seems like something that would be useful.

I do wonder if we should make passing in `core_numbers` optional, and if they are not passed then also call `core_numbers`.  This would allow a simple caller that just wants to extract a 3-core from the graph to be able to call the function directly and get the complete answer, while a more sophisticated case might call `core_number` once and reuse the result to extract the different subgraphs as required.

Reply:
Yeah... +1 for taking `core_numbers` as std::optional.

Still a bit not-sure about "Ultimately I concluded that if we're going to expose a complete k-core implementation at the python level we should probably expose it at the lower levels as well."

At the python layer, user convenience might be more important than anything else. And providing an additional python function that internally calls core_number, find vertex list, and induced subgraph has no toll on compile time/binary size as python is an interpreted language.

Not sure we should provide the same level of convenience for C++ users, as we assume C++ users are more advanced and they might be OK about finding k-core by composing core_numbers and induced subgraph.

But this is a much bigger topic, and we may have this discussion again in the future.

Reply:
And another thing to consider is that this may incur some sort of dependency within a same algorithm level in our C++ software hierarchy. Most algorithms are composed of primitives and thrust calls. But now algorithms are composed of primitives, thrust calls, and other algorithms. I am not sure what kind of complications this can incur in the future.

Composing C++ algorithms in the C layer might be OK, but I am not sure whether we should support use cases that can be supported simply by composing few existing algorithms in the C++ level.

Reply:
Fair point.

I will change `core_numbers` to be optional.

We can explore dropping the `k_core` from the C++ layer if having the algorithms depend on other algorithms becomes an issue.  I see the point that adding it at the C layer may be sufficient.